### PR TITLE
Use local game art

### PR DESCRIPTION
**Pull request by tomaz82**
try to set default game art ( image and video ) from romfolder/images if the gamelist didn't specify an image or video

### CODE REVIEW

**Comment by tomaz82** [comment]
Change the function to setDefaultGameArt since it doesn't really load anything

**Comment by jrassa** [comment]
I have had an item on my todo list to do something similar based on [this request](https://retropie.org.uk/forum/topic/11636/emulationstation-updates-and-guide/34) from the forums.
> Right now I'm trying to find an ES feature request post. Every rom launcher I've ever used has been able to load screenshots/box-art from a sub folder in the rom folder. So if the image name matched the game, it was displayed. I was wanting this added. It would reduce my boot time from 25 seconds down to 5 :D

I am OK with the implemented functionality, but would prefer that the command line argument was reversed (off by default, using argument to enable) to leave open the possibility to implemented the above requested functionality.

**Comment by tomaz82** [comment]
@jrassa You mean like this? Not sure what you meant by "to leave open the possibility to implemented the above requested functionality" ? Doesn't this implement the requested functionality?

**Comment by zigurana** [comment]
Hi @tomaz82 , thanks for this contribution.
I am wondering if this change is not more at home in ThemeData.cpp? I would not expect each and every SystemData to have knowledge of a default image, but rather that the theme has a way to provide a fallback in case there is no selection-specific image available. Why did you place this in SystemData.cpp?

On that note: [this recent commit](https://github.com/RetroPie/EmulationStation/commit/24f1b969bf258022304ecc059bb9e34f7d958b3e) by @jrassa already allows themes to supply a default image, how does this PR mesh with that addition?
I understand this PR gives the control to the user, via the command line parameter, rather than the theme-creator, but what is the added benefit of that? Would it not be better for the theme to be updated with a (theme specific) default?

Minor concern: I'm not a fan of the hard-coded `/image` path, can we not come up with a user-defined path?

**Comment by tomaz82** [comment]
@zigurana This doesn't load "a" default image, this checks for an image and video for each rom if the gamelist didn't specify one, this has nothing to do with theme data. So yeah, the wording "default" could be improved I guess.

**Comment by zigurana** [comment]
Ok, then maybe I am not understanding the use-case for this PR. 
If you are introducing a second location for items to look for metadata-assets (eg:  `startpath/images/[itemname]-image.png`), why not just populate the gamelist.xml entry?

**Comment by tomaz82** [comment]
The use-case is for when you have a bunch of roms, you find / download / make images and dvideos for them, you put them in ./images and fire up ES and voila, they're displayed. without having to manually edit the gamelist.xml or run a scraper.

Also see jrassas comment regarding the feature request from the forum.

**Comment by jrassa** [comment]
> @jrassa You mean like this? Not sure what you meant by "to leave open the possibility to implemented the above requested functionality" ? Doesn't this implement the requested functionality?

The difference is that this is populating the metadata which will then persist in the gamelist.xml files.  Functionally, it is almost like a local scrapper.  The request I referenced specifically noted that they did not want the paths stored in the gamelist.xml files to improve startup speed.

From an implementation standpoint, the difference would be to move the check into [updateInfoPanel](https://github.com/RetroPie/EmulationStation/blob/master/es-app/src/views/gamelist/DetailedGameListView.cpp#L197) on DetailedGamelistView/VideoGameListView

**Comment by tomaz82** [comment]
@jrassa Aahh! I see,  I can defeinitely change it to do that, eventho that look a lot less elegant codewise.

Should that be with, or without a setting to enable / disable it?

**Comment by tomaz82** [comment]
There, changed it to behave like what was asked for in the feature request.

**Comment by zigurana** [comment]
I like where this is going, good work!
However, I am going to be a bit difficult again, sorry (but I hope I can make a fair case here).

In your current proposal, you are putting the burden of figuring out where to get the correct image/video path on the view. I think this should be a responsibility of the FileData, which also has (owns) the metadata. I do not think the view needs to know where the correct path comes from, as long as it gets it.

So rather than complicating the Detailed/VideoView, I would really like to see:
1. An extended `[FileData::getVideoPath()](https://github.com/RetroPie/EmulationStation/blob/master/es-app/src/FileData.cpp#L79)`, that checks for a valid entry in the metadata and then tries the default location. The same holds for `FileData::getMarquee()`, `FileData::getThumbnail()`, and other assets that might be "hardwired" like this.
2. A new method `FileData::getImage()` because why is it not already there?? That should be the method being called by the [DetailedGameListView](https://github.com/RetroPie/EmulationStation/blob/master/es-app/src/views/gamelist/DetailedGameListView.cpp#L197), similar to what happens in the [VideoGameListView](https://github.com/RetroPie/EmulationStation/blob/master/es-app/src/views/gamelist/VideoGameListView.cpp#L246).

Further, a mention was made that this alternative approach to XML path resolution might offer a performance benefit, but I am sceptical that this is the case. In the current form, we a using both strategies, and there certainly is no performance benefit in doing that. 
For users that do not use a `gamelist.xml` at all, and run emulationstation with `"--ignore-gamelist"`, there might be a benefit, but I'd like to see some numbers to substantiate that. 

Finally, some usability questions that need answering: 
* What would a user expect when a game has valid entry in the XML, AND in the default path location? 
* Do we need to support both strategies at the same time? 
* Currently the fixed location is a fallback for the xml, but why not the other way around? I can see users mass scraping most of their collection, but wanting to override some games manually.

In any case, this new strategy offers an easy way for users to supply image and video assets for their games, if they are somehow unable/unwilling to use the gamelist.xml, and that is a good thing!

**Comment by tomaz82** [comment]
I expect the performance benefit is from not having a huge xml that needs to be parsed. But if that's the case I'd also want to ( later on ) look at what can be done to speed up the xml parser.

I propose that the XML has highest priority.

The funny thing is that this is my third version of this code, the first one ( didn't create a PR for that one ) did what you ask for,  set the path in getVideoPath and getThumbnailPath .. :P

But yes, I like that approach more and will change it to do that.

**Comment by jrassa** [comment]
> So rather than complicating the Detailed/VideoView, I would really like to see:
> 
> An extended [FileData::getVideoPath()](https://github.com/RetroPie/EmulationStation/blob/master/es-app/src/FileData.cpp#L79), that checks for a valid entry in the metadata and then tries the default location. The same holds for FileData::getMarquee(), FileData::getThumbnail(), and other assets that might be "hardwired" like this.
> A new method FileData::getImage() because why is it not already there?? That should be the method being called by the DetailedGameListView, similar to what happens in the VideoGameListView.

FileData does seem like the best location.  This will also ensure it works with the automatic view detection.


> Further, a mention was made that this alternative approach to XML path resolution might offer a performance benefit, but I am sceptical that this is the case. In the current form, we a using both strategies, and there certainly is no performance benefit in doing that.
> For users that do not use a gamelist.xml at all, and run emulationstation with "--ignore-gamelist", there might be a benefit, but I'd like to see some numbers to substantiate that.

I think the idea behind the perceived performance benefit is that it should improved startup time caused by the parsing of the gamelist whereas this change would defer checking for individual files until the point that they are needed.  I don't know how much of a difference it will really make, but Rookervik expected an improvement from 25 seconds to 5 seconds.

> What would a user expect when a game has valid entry in the XML, AND in the default path location?
> Do we need to support both strategies at the same time?
> Currently the fixed location is a fallback for the xml, but why not the other way around? I can see users mass scraping most of their collection, but wanting to override some games manually.

The way I thought about it is the fixed location would be the default and the gamelist would override it.  I think it would be confusing for a user to have a path in the gamelist, but not be used.

**Comment by tomaz82** [comment]
Updated per suggestions from @zigurana and @jrassa

**Comment by jrassa** [comment]
@tomaz82 I can't say for sure, but I believe that the thumbnail tag was originally added as part of the incomplete grid view implementation that exists within the code.

File: es-app/src/views/gamelist/VideoGameListView.cpp [hunk 1]
```diff
@@ -240,37 +240,15 @@ void VideoGameListView::updateInfoPanel()
...
 
 	}else{
-		std::string				video_path;
-		std::string				marquee_path;
-		std::string				thumbnail_path;
-		video_path 			= file->getVideoPath();
-		marquee_path 		= file->getMarqueePath();
-		thumbnail_path 		= file->getThumbnailPath();
-
-		if	(!video_path.empty() && (video_path[0] == '~'))
```
Comment (jrassa) [inline es-app/src/views/gamelist/VideoGameListView.cpp]: I don't feel strongly about it either way, but removing the handling of '~' may break things for some users.  Everything else looks good to me.  I will try to test it out tonight.

Comment (tomaz82) [reply es-app/src/views/gamelist/VideoGameListView.cpp]: I removed that because that is already handled by resolvePath in createFromXML

**Comment by jrassa** [comment]
@tomaz82 Tested it out and it is working as expected.  Thanks.

**Comment by joolswills** [comment]
I'm going to revert this from master - sorry @tomaz82.

It's a useful feature, but it currently adds too much overhead looking for the art. Can be revisited at some point - if it was to cache/or save the information in the gamelist, then it would be better as it would only be slower on the first load.

**Pull request merged**